### ISSUE TITLE
Ingress with ngnix results in 502 bad gateway

### ISSUE THREAD

**Issue by shashankram** [openservicemesh/osm#2930]
**Bug description**:
While testing ingress to the azure-voting app frontend (https://github.com/Azure-Samples/azure-voting-app-redis), ingress with k8s nginx-ingress results in a 502 bad gateway.

Response:
```
HTTP/1.1 502 Bad Gateway
Date: Thu, 18 Mar 2021 18:09:37 GMT
Content-Type: text/html
Content-Length: 173
Connection: keep-alive
x-envoy-upstream-service-time: 1
```

Envoy stats for request failures:
```
cluster.test/azure-vote-front-local.internal.upstream_rq_502: 16
cluster.test/azure-vote-front-local.internal.upstream_rq_5xx: 16
cluster.test/azure-vote-front-local.internal.upstream_rq_completed: 16
```

Envoy logs pertaining to the request:
```
[2021-03-18 17:55:55.695][13][debug][rbac] [source/extensions/filters/http/rbac/rbac_filter.cc:113] enforced allowed, matched policy rbac-for-route
[2021-03-18 17:55:55.695][13][trace][http] [source/common/http/filter_manager.cc:498] [C16][S6167152077661295004] decode headers called: filter=0x46c67f277e00 status=0
[2021-03-18 17:55:55.695][13][debug][router] [source/common/router/router.cc:425] [C16][S6167152077661295004] cluster 'test/azure-vote-front-local' match for URL '/'
[2021-03-18 17:55:55.695][13][debug][router] [source/common/router/router.cc:582] [C16][S6167152077661295004] router decoding headers:
':authority', '20.57.104.237'
':path', '/'
':method', 'HEAD'
':scheme', 'http'
'x-request-id', '1887ee9589398c81f469fbf69b70aebe'
'x-real-ip', '10.240.0.4'
'x-forwarded-for', '10.240.0.4'
'x-forwarded-host', '20.57.104.237'
'x-forwarded-port', '80'
'x-forwarded-proto', 'http'
'x-scheme', 'http'
'user-agent', 'curl/7.68.0'
'accept', '*/*'
'x-envoy-internal', 'true'
'x-envoy-expected-rq-timeout-ms', '15000'

[2021-03-18 17:55:55.695][13][debug][pool] [source/common/http/conn_pool_base.cc:79] queueing stream due to no available connections
[2021-03-18 17:55:55.695][13][debug][pool] [source/common/conn_pool/conn_pool_base.cc:106] creating a new connection
[2021-03-18 17:55:55.695][13][debug][client] [source/common/http/codec_client.cc:41] [C17] connecting
[2021-03-18 17:55:55.695][13][debug][connection] [source/common/network/connection_impl.cc:860] [C17] connecting to 0.0.0.0:80
[2021-03-18 17:55:55.696][13][debug][connection] [source/common/network/connection_impl.cc:876] [C17] connection in progress
[2021-03-18 17:55:55.696][13][trace][http] [source/common/http/filter_manager.cc:498] [C16][S6167152077661295004] decode headers called: filter=0x46c67f277d80 status=1
[2021-03-18 17:55:55.696][13][trace][misc] [source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x46c67f276d80 for 300000ms, min is 300000ms
[2021-03-18 17:55:55.696][13][trace][http] [source/common/http/http1/codec_impl.cc:620] [C16] parsed 274 bytes
[2021-03-18 17:55:55.696][13][trace][connection] [source/common/network/connection_impl.cc:548] [C16] socket event: 2
[2021-03-18 17:55:55.696][13][trace][connection] [source/common/network/connection_impl.cc:657] [C16] write ready
[2021-03-18 17:55:55.696][13][trace][connection] [source/common/network/connection_impl.cc:548] [C17] socket event: 2
[2021-03-18 17:55:55.696][13][trace][connection] [source/common/network/connection_impl.cc:657] [C17] write ready
[2021-03-18 17:55:55.696][13][debug][connection] [source/common/network/connection_impl.cc:666] [C17] connected
[2021-03-18 17:55:55.696][13][trace][connection] [source/common/network/connection_impl.cc:407] [C17] raising connection event 2
[2021-03-18 17:55:55.696][13][debug][client] [source/common/http/codec_client.cc:80] [C17] connected
[2021-03-18 17:55:55.696][13][debug][pool] [source/common/conn_pool/conn_pool_base.cc:225] [C17] attaching to next stream
[2021-03-18 17:55:55.696][13][debug][pool] [source/common/conn_pool/conn_pool_base.cc:130] [C17] creating stream
[2021-03-18 17:55:55.696][13][debug][router] [source/common/router/upstream_request.cc:354] [C16][S6167152077661295004] pool ready
[2021-03-18 17:55:55.696][13][trace][connection] [source/common/network/connection_impl.cc:465] [C17] writing 356 bytes, end_stream false

[2021-03-18 17:55:55.696][13][trace][connection] [source/common/network/connection_impl.cc:657] [C17] write ready
[2021-03-18 17:55:55.696][13][trace][connection] [source/common/network/raw_buffer_socket.cc:68] [C17] write returns: 356
[2021-03-18 17:55:55.696][13][trace][connection] [source/common/network/connection_impl.cc:548] [C17] socket event: 2
[2021-03-18 17:55:55.696][13][trace][connection] [source/common/network/connection_impl.cc:657] [C17] write ready
[2021-03-18 17:55:55.698][13][trace][connection] [source/common/network/connection_impl.cc:548] [C17] socket event: 3
[2021-03-18 17:55:55.698][13][trace][connection] [source/common/network/connection_impl.cc:657] [C17] write ready
[2021-03-18 17:55:55.698][13][trace][connection] [source/common/network/connection_impl.cc:586] [C17] read ready. dispatch_buffered_data=false
[2021-03-18 17:55:55.698][13][trace][connection] [source/common/network/raw_buffer_socket.cc:25] [C17] read returns: 157
[2021-03-18 17:55:55.698][13][trace][connection] [source/common/network/raw_buffer_socket.cc:39] [C17] read error: Resource temporarily unavailable
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:571] [C17] parsing 157 bytes
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:847] [C17] message begin
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:494] [C17] completed header: key=Server value=nginx/1.13.7
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:494] [C17] completed header: key=Date value=Thu, 18 Mar 2021 17:55:55 GMT
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:494] [C17] completed header: key=Content-Type value=text/html
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:494] [C17] completed header: key=Content-Length value=173
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:699] [C17] onHeadersCompleteBase
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:494] [C17] completed header: key=Connection value=keep-alive
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:1198] [C17] status_code 502
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:1208] [C17] Client: onHeadersComplete size=5
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:823] [C17] message complete
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:1288] [C17] message complete
[2021-03-18 17:55:55.698][13][debug][client] [source/common/http/codec_client.cc:112] [C17] response complete
[2021-03-18 17:55:55.698][13][trace][main] [source/common/event/dispatcher_impl.cc:213] item added to deferred deletion list (size=1)

[2021-03-18 17:55:55.698][13][debug][router] [source/common/router/router.cc:1174] [C16][S6167152077661295004] upstream headers complete: end_stream=true
[2021-03-18 17:55:55.698][13][trace][misc] [source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x46c67f276d80 for 300000ms, min is 300000ms
[2021-03-18 17:55:55.698][13][debug][http] [source/common/http/conn_manager_impl.cc:1484] [C16][S6167152077661295004] encoding headers via codec (end_stream=true):
':status', '502'
'server', 'envoy'
'date', 'Thu, 18 Mar 2021 17:55:55 GMT'
'content-type', 'text/html'
'content-length', '173'
'x-envoy-upstream-service-time', '2'
```

**Affected area (please mark with X where applicable)**:
- Install                    [ ]
- SMI Traffic Access Policy  [ ]
- SMI Traffic Specs Policy   [ ]
- SMI Traffic Split Policy   [ ]
- Permissive Traffic Policy  [ ]
- Ingress                    [X]
- Egress                     [ ]
- Envoy Control Plane        [ ]
- CLI Tool                   [ ]
- Metrics                    [ ]
- Certificate Management     [ ]
- Sidecar Injection          [ ]
- Logging                    [ ]
- Debugging                  [ ]
- Tests                      [ ]
- Demo                       [ ]
- CI System                  [ ]

**Expected behavior**:
Ingress to the voting app should not 502. Per the logs, the ingress route is correctly applied and being routed to `azure-vote-front-local` local cluster for the `azure-vote-front` service.

**Steps to reproduce the bug (as precisely as possible)**:
1. Install OSM v0.8, enable permissive mode

2. Deploy the azure voting app and expose it via Nginx ingress
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: azure-vote-back
spec:
  replicas: 1
  selector:
    matchLabels:
      app: azure-vote-back
  template:
    metadata:
      labels:
        app: azure-vote-back
    spec:
      nodeSelector:
        "beta.kubernetes.io/os": linux
      containers:
      - name: azure-vote-back
        image: mcr.microsoft.com/oss/bitnami/redis:6.0.8
        env:
        - name: ALLOW_EMPTY_PASSWORD
          value: "yes"
        ports:
        - containerPort: 6379
          name: redis
---
apiVersion: v1
kind: Service
metadata:
  name: azure-vote-back
spec:
  ports:
  - port: 6379
    name: tcp-6379
  selector:
    app: azure-vote-back
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: azure-vote-front
spec:
  replicas: 1
  selector:
    matchLabels:
      app: azure-vote-front
  strategy:
    rollingUpdate:
      maxSurge: 1
      maxUnavailable: 1
  minReadySeconds: 5
  template:
    metadata:
      labels:
        app: azure-vote-front
    spec:
      nodeSelector:
        "beta.kubernetes.io/os": linux
      containers:
      - name: azure-vote-front
        image: mcr.microsoft.com/azuredocs/azure-vote-front:v1
        ports:
        - containerPort: 80
        resources:
          requests:
            cpu: 250m
          limits:
            cpu: 500m
        env:
        - name: REDIS
          value: "azure-vote-back"
---
apiVersion: v1
kind: Service
metadata:
  name: azure-vote-front
spec:
  ports:
  - port: 80
    name: http-80
  selector:
    app: azure-vote-front
---
kind: Ingress
apiVersion: networking.k8s.io/v1beta1
metadata:
  name: azure-front-ingress
  annotations:
    nginx.ingress.kubernetes.io/use-regex: "true"
spec:
  ingressClassName: nginx
  rules:
    - http:
        paths:
          - path: /.*
            backend:
              serviceName: azure-vote-front
              servicePort: 80
```

**How was OSM installed?**:  `osm install`

**Anything else we need to know?**: n/a

**Environment**:
- OSM version (use `osm version`): v0.8.0
- 
- Kubernetes version (use `kubectl version`): 
```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.7", GitCommit:"1dd5338295409edcfff11505e7bb246f0d325d15", GitTreeState:"clean", BuildDate:"2021-01-13T13:23:52Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.10", GitCommit:"41d24ec9c736cf0bdb0de3549d30c676e98eebaf", GitTreeState:"clean", BuildDate:"2021-01-18T09:12:27Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by draychev** [comment]
@shashankram could you share the Kubernetes Ingress resource you are using?

Can you share these:
1. Service
2. Pods for the Service
3. Endpoints for the Service
4. Ingress resource that references the Service

**Comment by shashankram** [comment]
> @shashankram could you share the Kubernetes Ingress resource you are using?
> 
> Can you share these:
> 
> 1. Service
> 2. Pods for the Service
> 3. Endpoints for the Service
> 4. Ingress resource that references the Service

I updated the description.

**Comment by draychev** [comment]
@shashankram can you post the SMI CRDs as well?

**Comment by shashankram** [comment]
> @shashankram can you post the SMI CRDs as well?

In my testing, the default CRDs embedded in the v0.8.0 osm cli are used: https://github.com/openservicemesh/osm/tree/release-v0.8/charts/osm/crds

**Comment by draychev** [comment]
@shashankram I was interested in the `TrafficTarget` and `TCPRoute`, but I realize this was in permissive-mode.

The 502 Error may be coming from 2 Nginxs
  1. the first one is in the `azure-vote-front` app. I saw a 502 there when the `front` app can't connect to Redis. This may happen for a few different reasons:
      1) the bug we've been chasing
      2) a matter of timing of Envoy configuration arriving
      3) no policies applied
  2. The cluster's Nginx / Ingress
  
To address the first Nginx (inside the front app) issue I added a health probe:
```
        readinessProbe:
          httpGet:
            path: /
            port: 80
            scheme: HTTP
          initialDelaySeconds: 5
          periodSeconds: 5
```
This restarts the pod if the pod can't connect to redis.

For the Ingress I crafter a slightly different Ingress YAML to workaround some of the deficiencies of the current alpha OSM Ingress implementation:
```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: azure-vote-ingress
  namespace: ${NAMESPACE}-front
  annotations:
    kubernetes.io/ingress.class: nginx
spec:
  rules:
    - host: azure-vote.contoso.com
      http:
        paths:
        - path: /
          backend:
            serviceName: azure-vote-front
            servicePort: 80
  backend:
    serviceName: azure-vote-front
    servicePort: 80
```

With all of these I was able to get to the app, while in the mesh, through a public IP address:
```bash
$ kubectl get ingress -n azure-vote azure-vote-ingress -o json | jq -r '.status.loadBalancer.ingress[0].ip'
xxx.137.99.xxx

$ curl -H 'Host: azure-vote.contoso.com' xxx.137.99.xxx
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
    <link rel="stylesheet" type="text/css" href="/static/default.css">
    <title>Azure Voting App</title>
...

$ kc get pods -n azure-vote

NAME                                READY   STATUS    RESTARTS   AGE
azure-vote-back-56d7bb6b6d-jf9gn    2/2     Running   0          46m
azure-vote-front-5666cdfb76-jql6s   2/2     Running   0          46m
```


I made a script w/ this demo here (will move it to the OSM repo): https://gist.github.com/draychev/787e2b9f63e99dba0f0290351825f89c

**Comment by shashankram** [comment]
@draychev Thanks for looking into this. 

When I first tried this, I noticed the ingress traffic making it all the till being routed to the local service cluster configured in `RDS_Inbound`, but not making it through thereafter:
```[2021-03-18 17:55:55.696][13][trace][connection] [source/common/network/connection_impl.cc:657] [C17] write ready
[2021-03-18 17:55:55.696][13][trace][connection] [source/common/network/raw_buffer_socket.cc:68] [C17] write returns: 356
[2021-03-18 17:55:55.696][13][trace][connection] [source/common/network/connection_impl.cc:548] [C17] socket event: 2
[2021-03-18 17:55:55.696][13][trace][connection] [source/common/network/connection_impl.cc:657] [C17] write ready
[2021-03-18 17:55:55.698][13][trace][connection] [source/common/network/connection_impl.cc:548] [C17] socket event: 3
[2021-03-18 17:55:55.698][13][trace][connection] [source/common/network/connection_impl.cc:657] [C17] write ready
[2021-03-18 17:55:55.698][13][trace][connection] [source/common/network/connection_impl.cc:586] [C17] read ready. dispatch_buffered_data=false
[2021-03-18 17:55:55.698][13][trace][connection] [source/common/network/raw_buffer_socket.cc:25] [C17] read returns: 157
[2021-03-18 17:55:55.698][13][trace][connection] [source/common/network/raw_buffer_socket.cc:39] [C17] read error: Resource temporarily unavailable
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:571] [C17] parsing 157 bytes
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:847] [C17] message begin
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:494] [C17] completed header: key=Server value=nginx/1.13.7
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:494] [C17] completed header: key=Date value=Thu, 18 Mar 2021 17:55:55 GMT
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:494] [C17] completed header: key=Content-Type value=text/html
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:494] [C17] completed header: key=Content-Length value=173
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:699] [C17] onHeadersCompleteBase
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:494] [C17] completed header: key=Connection value=keep-alive
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:1198] [C17] status_code 502
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:1208] [C17] Client: onHeadersComplete size=5
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:823] [C17] message complete
[2021-03-18 17:55:55.698][13][trace][http] [source/common/http/http1/codec_impl.cc:1288] [C17] message complete
[2021-03-18 17:55:55.698][13][debug][client] [source/common/http/codec_client.cc:112] [C17] response complete
[2021-03-18 17:55:55.698][13][trace][main] [source/common/event/dispatcher_impl.cc:213] item added to deferred deletion list (size=1)

[2021-03-18 17:55:55.698][13][debug][router] [source/common/router/router.cc:1174] [C16][S6167152077661295004] upstream headers complete: end_stream=true
[2021-03-18 17:55:55.698][13][trace][misc] [source/common/event/scaled_range_timer_manager_impl.cc:60] enableTimer called on 0x46c67f276d80 for 300000ms, min is 300000ms
[2021-03-18 17:55:55.698][13][debug][http] [source/common/http/conn_manager_impl.cc:1484] [C16][S6167152077661295004] encoding headers via codec (end_stream=true):
':status', '502'
'server', 'envoy'
'date', 'Thu, 18 Mar 2021 17:55:55 GMT'
'content-type', 'text/html'
'content-length', '173'
'x-envoy-upstream-service-time', '2'
```

It seems like the same symptom is not consistent across runs.

**Comment by shashankram** [comment]
I can't reproduce this issue with v0.8.2, so I am going to close this one. Note that there were some important bug fixes that were made available in v0.8.2 and the original bug was filed against v0.8.0, which we know had known issues related to SDS secrets.
